library/objective-c/EnvoyHTTPStreamImpl.m

```diff
@@ -137,27 +137,25 @@ static void ios_on_complete(void *context) {
   EnvoyHTTPCallbacks *callbacks = c->callbacks;
   EnvoyHTTPStreamImpl *stream = c->stream;
   dispatch_async(callbacks.dispatchQueue, ^{
-    if (atomic_load(c->canceled)) {
-      return;
-    }
-
+    // FIXME: If the callback queue is not serial, clean up is not currently thread-safe.
     assert(stream);
     [stream cleanUp];
   });
 }
 
 static void ios_on_cancel(void *context) {
+  // This call is atomically gated at the call-site and will only happen once.
```

Comment:
I filed an issue to add a flowchart for this so behavior is clearer: https://github.com/lyft/envoy-mobile/issues/584

Reply:
It'd also be good if we could outline here that `onCancel()` can be propogated through callbacks even after a stream is technically completed